File: BosonJobMaster.R
# BosonJobMaster.R
# run: Rscript BosonJobMaster.R 0 bootstrap-r-jobs 2 0 0 s3://boson-base/rboson-test/

# run: Rscript BosonJobMaster.R 0 run-r-tasks 1 1 1 s3://boson-base/rboson-test/
# run: Rscript BosonJobMaster.R 0 run-r-tasks 1 1 1,2,3,4,5 s3://boson-base/rboson-test/

# load libraries & scripts
source('s3utils.R')
source('AWSBatchUtils.R')

# parse arguments
args = commandArgs(TRUE)
print('Reading job arguments:'); print(args)
if (length(args) == 6) {
  batch.id  = as.character(args[1])
  job.type  = as.character(args[2])
  njobs     = as.integer(args[3])
  job.id    = as.integer(args[4])
  task.ids  = as.integer(strsplit(args[5], ',')[[1]])
  s3.path   = as.character(args[6])
} else {
  stop("Need to pass 6 arguments")
}

# get metadata and load in environment
print('Loading inputs:')
meta = LoadObjectsFromS3(
  s3.path = s3.path,
  keys = c(paste0('batch_', batch.id, '_in'))
)
list2env(meta, envir = environment())

if (job.type == 'bootstrap-r-jobs') {
  
  # bootstrap jobs in the batch
  df.jobid = BootstrapBatchJobs (
    batch.id = batch.id,
    ntasks = length(X),
    njobs = njobs,
    s3.path = s3.path
  )
  print(df.jobid)
  
  # save outcome in S3
  SaveObjectesInS3(
    out = list(df.jobid = df.jobid),
    s3.path = s3.path,
    key = c(paste0('batch_', batch.id, '_jobids'))
  )
  
} else if (job.type == 'run-r-tasks') {
  
  # run assigned jobs
  out = list()
  if (!exists('extra.args')) {
    extra.args = list()
  }
  
  for (i in task.ids) {
    cat(paste('** Starting job', i, '**\n'))
    tryCatch({
      out[[length(out)+1]] = do.call(FUN, append(X[[i]], extra.args))
      names(out)[length(out)] = names(X)[i]
    }, error = function(e) {
      cat(paste('Error on task', i, '\n'))
    },
    finally = cat(paste('Completed task', i, '\n\n'))
    )
  }
  
  # save outcome in S3
  SaveObjectesInS3(
    out = out,
    s3.path = s3.path,
    key = c(paste0('batch_', batch.id, '_out_', job.id))
  )
  
}

File: BosonJobMaster.py
# run: python BosonJobMaster.py 1 bootstrap-py-jobs 4 0 0 s3://boson-base/pyboson-test/
# run: python BosonJobMaster.py 1 run-py-tasks 4 1 1,2,3,4 s3://boson-base/pyboson-test/

import sys
import numpy as np

from s3utils import * 
from AWSBatchUtils import *

# parse arguments
arguments = sys.argv[1:]
print 'Reading job arguments:\n {}'.format(arguments)

if (len(arguments) == 6):
	batch_id  = arguments[0]
	job_type  = arguments[1]
	njobs     = int(arguments[2])
	job_id    = int(arguments[3])
	task_ids  = map(int, arguments[4].split(","))
	s3_path   = arguments[5]
	print batch_id, job_type, njobs, job_id, task_ids, s3_path 
else :
	print "Need to pass 6 arguments"
	sys.exit()

# get metadata and load in environment
print('Loading inputs:')
meta = LoadObjectsFromS3(s3_path = s3_path, 
	keys = ['batch_{}_in'.format(batch_id),])[0]
FUN = meta['FUN']
X   = meta['X']
extra_args = {key: meta[key] for key in meta.keys() if key not in ['FUN', 'X']}

if (job_type == 'bootstrap-py-jobs') :
	## bootstrap py-jobs ##
	num_tasks = len(X)
	if (num_tasks == 1) :
		task_partitions = [[1,],]
	else :
		tasks = np.arange(num_tasks)
		task_partitions = np.split(tasks, njobs)

	out = {}
	job_idx = 0

	for t in task_partitions :
		job_idx += 1
		task_ids = ','.join(map(str, t))
		print 'Submitting tasks: {}'.format(task_ids)
		out[str(job_idx)] = SubmitBatchJobs(
			batch_id = batch_id,
			njobs = 1,
			s3_path = s3_path,
			job_type = 'run-py-tasks',
			job_id = str(job_idx),
			task_ids = task_ids)

	SaveObjectesInS3(
		s3_path = s3_path,
		key = 'batch_{}_jobids'.format(batch_id),
		out = out)

elif (job_type == 'run-py-tasks'):
	# run py-jobs 
	# run assigned jobs
	out = {}

	for i in task_ids :
		print '** Starting job{} **\n'.format(i)
		try:
			all_args = dict(X[i], **extra_args)
			out[str(i)] = FUN(**all_args)
		except:
			print('error in task' + str(i))

  	# save outcome in S3
  	SaveObjectesInS3(
  		s3_path = s3_path,
  		key = 'batch_{}_out_{}'.format(batch_id,job_id),
    	out = out)